project/src/main/java/com/google/sps/data/Election.java

```diff
@@ -14,4 +14,56 @@
 
 package com.google.sps.data;
 
-public class Election {}
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class Election {
+  private String name;
+  private Date date;
+  private List<Position> candidatePositions;
+
+  public Election(String inputName, Date inputDate, List<String> inputCandidateIds,
+      List<String> inputCandidateNames, List<String> inputCandidatePositions,
+      List<String> inputCandidatePartyAffiliation, List<String> inputCandidateIncumbency) {
+    this.name = inputName;
+    this.date = inputDate;
+    // Reformat database data to correlate one {@code Position} object with a list of candidates
+    // and their information.
+    Set<String> distinctPositions = new HashSet<>(inputCandidatePositions);
+    this.candidatePositions = new ArrayList<>(distinctPositions.size());
+    for (String position : distinctPositions) {
+      int startIndex = inputCandidatePositions.indexOf(position);
+      int endIndex = inputCandidatePositions.lastIndexOf(position);
+      Position candidatePosition = new Position(position,
```

Comment:
TBH I'm not a fan of this design where you're assuming  the indices correspond to the same candidate across the different lists. Have you thought about having a map instead? If not, feel free to reply with your logic for going with this approach :)

Reply:
Yes, I agree that this is not ideal. I thought about having a map. The reason why I used this approach is because manipulating the database to keep an object (election) of object (positions) of object (candidates running for the positions) is complex. I researched into this, and this is doable via Datastore ancestor paths. Alternatively, I can also achieve this by creating an intermediary table for position VS candidates. However, the trade-off would be complexities with (1) correctly formatting everything into the database and (2) querying the database.
Because of these reasons, I decided to go with the easiest approach. But I'm totally open to changing it

Reply:
Yes, candidates can run for different positions or be part of different elections. But candidate information won't be duplicated. Currently there is a table of candidates, keyed by ID. There is a table of elections, keyed by election names. Each election entry has a list of positions and a list of candidate IDs. The i-th candidate is running for the i-th position.

Reply:
Should I create a separate table of position-candidate ID pairs (or say position objects)? Then each election would only contain a list of position IDs. Each position entry would contain a candidate ID, and whether the candidate is an incumbent.

Reply:
Oh does the database not easily store object lists?

Reply:
I think the thing that most feels weird is the 4 lists of strings with candidate information instead of 1 list of "candidate" objects.

Reply:
Yeah the database doesn't easily store list of objects. Lists of IDs are easy though.

Oh I see what you mean by duplication now. Apologies for the confusion, the 4 lists of strings are a brief set of candidate data that are organized in a format that will be easily parsed by the frontend. It is possible that one candidate can appear multiple times. The "id" and "name" property may be duplicated if a candidate runs for multiple positions/elections, while the incumbency is specific to the position. In the database, there is only a reference to the candidate ID, so information is not duplicated in the database.

Reply:
Would you suggest I instead assemble data in the frontend? Then the frontend would search for candidate information based on ID and then organize the data.

Reply:
Ah good to know that the candidate info isn't duplicated in the database.  This class is mostly for packaging the information for passing to the front end, correct?  In that case would it be valid to have a structure like:
```
Election(inputName, inputDate, List<Position> inputPositions)
Position(inputName, List<Candidate> candidates)
Candidate(id, name, partyAffiliation, currentPositionId)
```
And actually construct the Candidate and Position objects in DataServlet.java to pass into the Election object?

`Candidate` could also be called something like `DirectoryCandidate` so you know it's only the info for the directory page...or you could just reuse the Candidate object and only populate certain fields if enough overlaps idk

Reply:
Yes, this class is for packaging information for the front end.

Got it, I'll revise the structure. I'll rename `Candidate` to `DirectoryCandidate` for now.

Would you recommend modifying the database schema as well and create a Position type?

Reply:
No need to modify the database schema as long as the data is being stored and can be transformed to this form relatively efficiently.

Reply:
Got it, thanks! I'll start making the changes.